### PR TITLE
[Merged by Bors] - feat: port Algebra.Module.Pi

### PR DESCRIPTION
**Pull request by hrmacbeth**
One earlier failure was extensively discussed:
- https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Performance.20issue.20with.20.60CompleteBooleanAlgebra.60/near/319019205
- https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/type.20class.20inference.20looping

and is fixed as of https://github.com/leanprover/lean4/commit/70a6c06eef060d8e14869e2848749287e5364742, in mathlib as of the bump #1335.

Another failure posted to Zulip and the Lean 4 repo
- https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/type.20class.20inference.20failure.20in.20pi.20type
- https://github.com/leanprover/lean4/issues/2011

and is fixed as of https://github.com/leanprover/lean4/commit/fedf235cba35ed8bf6bf571cf38e6d8536b904ac, in mathlib as of the bump #1397.

There is one more mysterious `apply` failure, now worked around; we should track this down someday.

- [x] depends on: #1397

### CODE REVIEW

**Comment by hrmacbeth** [comment]
One of the errors will be fixed after https://github.com/leanprover/lean4/commit/70a6c06eef060d8e14869e2848749287e5364742
makes it downstream to mathlib.  Might as well wait and see if all of them are.

[Zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/type.20class.20inference.20looping)

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#1397
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#1397~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Algebra/Module/Pi.lean [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  { Pi.distribMulAction _ with
+    add_smul := fun _ _ _ => funext fun _ => add_smul _ _ _
+    zero_smul := fun _ => funext fun _ => zero_smul α _ }
+#align pi.module Pi.module
+
+/- Extra instance to short-circuit type class resolution.
+For unknown reasons, this is necessary for certain inference problems. E.g., for this to succeed:
+```lean
+example (β X : Type*) [normed_add_comm_group β] [normed_space ℝ β] : module ℝ (X → β) :=
+infer_instance
```
Comment (jcommelin) [inline Mathlib/Algebra/Module/Pi.lean:77]: Please Lean4ify

File: Mathlib/Algebra/Module/Pi.lean [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+```
+See: https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Typeclass.20resolution.20under.20binders/near/281296989
+-/
+/-- A special case of `pi.module` for non-dependent types. Lean struggles to elaborate
```
Comment (jcommelin) [inline Mathlib/Algebra/Module/Pi.lean:81]:
```suggestion
/-- A special case of `Pi.module` for non-dependent types. Lean struggles to elaborate
```

**Review by jcommelin — APPROVED** [review]
bors d+

**Pull request closed without merging**